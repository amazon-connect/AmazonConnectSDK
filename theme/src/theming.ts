import { applyTheme, Theme } from "@cloudscape-design/components/theming";

const connectTeal600 = "#077398";
const connectTeal700 = "#065B78";

const theme: Theme = {
  tokens: {
    colorBackgroundButtonPrimaryDefault: connectTeal600,
    colorBackgroundButtonPrimaryActive: connectTeal700,
    colorBackgroundButtonPrimaryHover: connectTeal700,

    colorBackgroundControlChecked: connectTeal600,
    // TODO waiting for latest public version
    // colorBackgroundSegmentActive: connectTeal600,
    // Not themable
    // colorBackgroundProgressBarContentDefault: connectTeal600,

    colorBorderButtonNormalDefault: connectTeal600,
    colorBorderButtonNormalHover: connectTeal700,

    colorTextLinkDefault: connectTeal600,
    colorTextLinkHover: connectTeal700,
    // TODO waiting for latest public version
    // colorTextSegmentHover: connectTeal600,

    colorTextButtonNormalDefault: connectTeal600,
    colorTextButtonNormalHover: connectTeal700,

    colorBorderItemSelected: connectTeal600,
    // TODO contrast super low
    colorBorderItemFocused: connectTeal600,
    // TODO what should this be?
    // colorBackgroundItemSelected: connectTeal600,

    colorTextAccent: {
      light: connectTeal600,
      dark: connectTeal600,
    },
  },
};

// Preparing the public signature for when we support overrides, but typing it so it's not usable for now.
export function applyConnectTheme(overrides?: never) {
  if (overrides) {
    throw new Error(
      "Connect theming for Cloudscape does not support overrides yet",
    );
  } else {
    applyTheme({ theme });
  }
}
