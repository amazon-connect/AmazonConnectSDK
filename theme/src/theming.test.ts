import { applyTheme } from "@cloudscape-design/components/theming";

import { applyConnectTheme } from "./theming";

jest.mock("@cloudscape-design/components/theming");
const applyThemeMock = applyTheme as jest.MockedFn<typeof applyTheme>;

// Not deleting this as it'll come back as soon as the theme override API is stable
// function expectTokens(tokens: Theme["tokens"]) {
//   expect(applyThemeMock).toHaveBeenCalledWith({
//     theme: {
//       // Need to cast as unknown because this whole repo uses @typescript-eslint/no-unsafe-assignment
//       // which produces false positives like this one. The typical recommendation is to instead rely on
//       // noImplicitAny from the Typescript compiler and no-explicit-any from ESLint.
//       tokens: expect.objectContaining(tokens) as unknown,
//     },
//   });
// }

describe("theming", () => {
  beforeEach(() => applyThemeMock.mockReset());

  it("applies default Connect theme", () => {
    applyConnectTheme();
    expect(applyThemeMock).toHaveBeenCalledTimes(1);
    // This is the only use-case that ever makes sense for snapshot tests.
    // We need an exact, precise output, and we want changes to it to be reviewed carefully.
    expect(applyThemeMock.mock.lastCall?.[0]).toMatchSnapshot();
  });

  // Not deleting these tests as they'll come back as soon as the theme override API is stable
  // it("allows non-modal overrides", () => {
  //   applyConnectTheme({ fontFamily: "Comic Sans", brandColor: "hotpink" });
  //   expectTokens({
  //     fontFamilyBase: "Comic Sans",
  //     colorBackgroundButtonPrimaryDefault: "hotpink",
  //     colorBackgroundButtonPrimaryActive: {
  //       dark: "#A7C1D1",
  //       light: "#065B78",
  //     },
  //     colorBackgroundButtonPrimaryHover: {
  //       dark: "#A7C1D1",
  //       light: "#065B78",
  //     },
  //   });
  // });
  //
  // it("allows modal overrides", () => {
  //   applyConnectTheme({ brandColor: { light: "deeppink", dark: "hotpink" } });
  //   expectTokens({
  //     colorBackgroundButtonPrimaryDefault: {
  //       light: "deeppink",
  //       dark: "hotpink",
  //     },
  //     colorBackgroundButtonPrimaryActive: {
  //       dark: "#A7C1D1",
  //       light: "#065B78",
  //     },
  //     colorBackgroundButtonPrimaryHover: {
  //       dark: "#A7C1D1",
  //       light: "#065B78",
  //     },
  //   });
  // });
  //
  // it("ignores unknown tokens", () => {
  //   applyConnectTheme();
  //   const defaultTheme = applyThemeMock.mock.lastCall;
  //   // @ts-expect-error - We use a property that doesn't exist on purpose to mimic non-TS customers.
  //   applyConnectTheme({ pageBorderRadius: "0" });
  //   expect(applyThemeMock.mock.lastCall).toEqual(defaultTheme);
  // });
  //
  // it("ignores modal overrides on non-modal properties", () => {
  //   applyConnectTheme();
  //   const defaultTheme = applyThemeMock.mock.lastCall;
  //   // @ts-expect-error - We use the wrong value type for fontFamily to mimic non-TS customers.
  //   applyConnectTheme({ fontFamily: { light: "deeppink", dark: "hotpink" } });
  //   expect(applyThemeMock.mock.lastCall).toEqual(defaultTheme);
  // });
});
