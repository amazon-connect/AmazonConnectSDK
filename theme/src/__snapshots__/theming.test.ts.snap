// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`theming applies default Connect theme 1`] = `
{
  "theme": {
    "tokens": {
      "colorBackgroundButtonPrimaryActive": {
        "dark": "#A7C1D1",
        "light": "#065B78",
      },
      "colorBackgroundButtonPrimaryDefault": {
        "dark": "#7CA2BB",
        "light": "#077398",
      },
      "colorBackgroundButtonPrimaryHover": {
        "dark": "#A7C1D1",
        "light": "#065B78",
      },
      "colorBackgroundControlChecked": {
        "dark": "#7CA2BB",
        "light": "#077398",
      },
      "colorBackgroundSegmentActive": {
        "dark": "#7CA2BB",
        "light": "#077398",
      },
      "colorBorderButtonNormalActive": {
        "dark": "#A7C1D1",
        "light": "#065B78",
      },
      "colorBorderButtonNormalDefault": {
        "dark": "#7CA2BB",
        "light": "#077398",
      },
      "colorBorderButtonNormalHover": {
        "dark": "#A7C1D1",
        "light": "#065B78",
      },
      "colorBorderItemFocused": {
        "dark": "#7CA2BB",
        "light": "#077398",
      },
      "colorBorderItemSelected": {
        "dark": "#7CA2BB",
        "light": "#077398",
      },
      "colorTextAccent": {
        "dark": "#7CA2BB",
        "light": "#077398",
      },
      "colorTextButtonNormalActive": {
        "dark": "#A7C1D1",
        "light": "#065B78",
      },
      "colorTextButtonNormalDefault": {
        "dark": "#7CA2BB",
        "light": "#077398",
      },
      "colorTextButtonNormalHover": {
        "dark": "#A7C1D1",
        "light": "#065B78",
      },
      "colorTextLinkDefault": {
        "dark": "#7CA2BB",
        "light": "#077398",
      },
      "colorTextLinkHover": {
        "dark": "#A7C1D1",
        "light": "#065B78",
      },
      "colorTextSegmentHover": {
        "dark": "#7CA2BB",
        "light": "#077398",
      },
      "fontFamilyBase": undefined,
    },
  },
}
`;
